### PR TITLE
feat: introduce controlled switchover of a replica cluster

### LINKED ISSUE #3790: [Feature]: Add `shutdownCheckpointToken` in the status after fencing/hibernating
### Is there an existing issue already for this feature request/idea?

- [X] I have searched for an existing issue, and could not find anything. I believe this is a new feature request to be evaluated.

### What problem is this feature going to solve? Why should it be added?

When fencing or hibernating a Postgres cluster, it would be handy to add in the status a token containing information about write-ahead logging and checkpoint processing.

### Describe the solution you'd like

We should add a new field in the status called `shutdownCheckpointToken` that is normally empty.

Following a fencing operation (and consequently hibernation too), `shutdownCheckpointToken` is an encoded string containing the following information from `pg_controldata`:

- Database system identifier
- Latest checkpoint's TimeLineID
- Latest checkpoint's REDO location
- Latest checkpoint's REDO WAL file
- Time of latest checkpoint

The above values must be taken when the `Database cluster state` is `shut down`.

### Describe alternatives you've considered

N/A

### Additional context

N/A

### Backport?

Yes

### Are you willing to actively contribute to this feature?

Yes

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### LINKED ISSUE #3944: [Feature]: Add `replica.token` option
### Is there an existing issue already for this feature request/idea?

- [X] I have searched for an existing issue, and could not find anything. I believe this is a new feature request to be evaluated.

### What problem is this feature going to solve? Why should it be added?

We do not have a way to ensure that a former replica cluster is promoted as a primary without any data loss.

### Describe the solution you'd like

We should introduce a new option called `replica.token` that is evaluated by a replica cluster when it transitions from replica to primary mode. The token format is the one described in #3790, and contains some fundamental information, including:

- Database system identifier
- Latest checkpoint's TimeLineID
- Latest checkpoint's REDO location
- Latest checkpoint's REDO WAL file
- Time of latest checkpoint

Consider this simplified scenario with `pg2` being a replica cluster of `pg1`.

`pg1` looks like this:

```apiVersion: postgresql.cnpg.io/v1
kind: Cluster
metadata:
  name: pg1
spec:
   # ...
  replica:
    primary: pg1
    source: pg2
    enabled: auto
```

and `pg2`:

```apiVersion: postgresql.cnpg.io/v1
kind: Cluster
metadata:
  name: pg2
spec:
   # ...
  replica:
    primary: pg1
    source: pg1
    enabled: auto
```

We change the `pg1` cluster, informing it that the new distributed primary is `pg2`:

```
  # pg1 cluster
  replica:
    primary: pg2
    source: pg2
```

This operation cleanly shuts down `pg1` and adds `shutdownCheckpointToken` to its status.

We grab the token and modify `pg2` as follows:

```
  # pg2 cluster
  replica:
    primary: pg2
    source: pg1
    token: <SHUTDOWN_CHECKPOINT_TOKEN_FROM_PG1>
```

The operator detects that the `primary` in `pg2` has changed, and that there's a token. We extract the information from the token and verify that we are on the same timeline and not ahead of the LSN. If so, we set on the designated primary:

- `recovery_target_action` to `promote` (we need to double check this during implementation)
- `recovery_target_timeline` to the one in the token
- `recovery_target_lsn` to the one in the token

We need to decide what to do in case somebody changes the `token` field and it is not compatible with the current timeline.

The `token` field represents the last time the current cluster was requested to be promoted.

The patch should also include a command for the plugin to facilitate this operation. We should also consider expanding the information we have in the token, if it helps us improve the resiliency of the operation.

### Describe alternatives you've considered

We have thought about using a table in the database to control this, but discarded as we prefer to use configuration.

### Additional context

This ticket depends on the following issues:

- #3790 
- #3796 
- #3942 

### Backport?

No

### Are you willing to actively contribute to this feature?

Yes

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### LINKED ISSUE #4586: [Feature]: Archive shutdown checkpoint WAL file after demotion
### Is there an existing issue already for this bug?

- [X] I have searched for an existing issue, and could not find anything. I believe this is a new bug.

### I have read the troubleshooting guide

- [X] I have read the troubleshooting guide and I think this is a new bug.

### I am running a supported version of CloudNativePG

- [X] I have read the troubleshooting guide and I think this is a new bug.

### Contact Details

_No response_

### Version

1.23.1

### What version of Kubernetes are you using?

1.29

### What is your Kubernetes environment?

Cloud: Amazon EKS

### How did you install the operator?

YAML manifest

### What happened?

With 1.23.1, when demoting the primary, the WAL file following the shutdown checkpoint wasn't archived:

```shell
kubectl exec jeeg-eu-central-1-1 -- pg_controldata|grep REDO
Defaulted container "postgres" out of: postgres, bootstrap-controller (init)
Latest checkpoint's REDO location:    0/24000028
Latest checkpoint's REDO WAL file:    000000010000000000000009
```

Log reports:

```
{"level":"info","ts":"2024-05-17T15:54:51Z","logger":"wal-archive","msg":"Archived WAL file","logging_pod":"jeeg-eu-central-1-1","walName":"pg_wal/000000010000000000000008","startTime":"2024-05-17T15:54:50Z","endTime":"2024-05-17T15:54:51Z","elapsedWalTime":1.093924041}
{"level":"info","ts":"2024-05-17T15:56:46Z","logger":"postgres","msg":"record","logging_pod":"jeeg-eu-central-1-1","record":{"log_time":"2024-05-17 15:56:46.897 UTC","process_id":"20","session_id":"66477c68.14","session_line_num":"14","session_start_time":"2024-05-17 15:48:56 UTC","transaction_id":"0","error_severity":"LOG","sql_state_code":"55P02","message":"parameter \"archive_mode\" cannot be changed without restarting the server","backend_type":"postmaster","query_id":"0"}}
```

Steps to reproduce are detailed here: https://github.com/gbartolini/postgres-kubernetes-playground/tree/main/aws-eks/examples/jeeg#simulating-a-data-center-switchover

This prevents the other cluster from being promoted without a gap.

### Cluster resource

_No response_

### Relevant log output

```shell
See above
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by armru**
Introduce support for the controlled demotion of a primary cluster and the subsequent promotion of a replica cluster (typically in another region) without the need to re-clone the former primary.

This operation will necessarily result in a downtime, which is typically scheduled and expected.

Demotion of a primary cluster is currently triggered by changing `replica.enabled` from `false` to `true` and includes:

- Shutting down the PostgreSQL server.
- Performing an extraordinary archival of the checkpoint REDO WAL file, containing the shutdown checkpoint, as a `.partial` file in the WAL archive.
- Generating the `demotionToken` in the `Cluster` status, containing information in base64-encoded JSON format retrieved from `pg_controldata` (this includes the database system identifier, latest checkpoint's `TimeLineID`, latest checkpoint's REDO location, latest checkpoint's REDO WAL file, and the time of the latest checkpoint).

Currently, promoting a replica cluster to a primary cluster is triggered by simultaneously:

- Adding the `replica.promotionToken` and setting its value to the `demotionToken` returned by the former primary cluster.
- Changing `replica.enabled` from `true` to `false`.

The patch includes validating the submitted token (it is the user's responsibility to set the correct value to ensure self-healing).

After waiting for the designated primary to replicate all the WAL information until the desired LSN (contained in the token) is reached, the promotion process is started.

Closes #3790 
Closes #3944   
Closes #4586

### CODE REVIEW

File: api/v1/cluster_types.go [hunk 1]
```diff
@@ -923,6 +923,15 @@ type ClusterStatus struct {
 	// SwitchReplicaClusterStatus is the status of the switch to replica cluster
 	// +optional
 	SwitchReplicaClusterStatus SwitchReplicaClusterStatus `json:"switchReplicaClusterStatus,omitempty"`
+
+	// ShutdownCheckpointToken is a json token containing the following information from pg_controldata:
+	//	Database system identifier
+	//	Latest checkpoint's TimeLineID
+	//	Latest checkpoint's REDO location
+	//	Latest checkpoint's REDO WAL file
+	//	Time of latest checkpoint
```
Comment (gbartolini) [inline api/v1/cluster_types.go:932]:
```suggestion
	// shutdownCheckpointToken is a JSON token containing the information
	// from pg_controldata such as Database system identifier, Latest checkpoint's
	// TimeLineID, Latest checkpoint's REDO location, Latest checkpoint's REDO
	// WAL file, and Time of latest checkpoint
```

**Comment by leonardoce** [comment]
E2e: https://github.com/EnterpriseDB/cloudnative-pg/actions/runs/9566577472

**Comment by leonardoce** [comment]
E2e tests are fine. I started another run here: https://github.com/EnterpriseDB/cloudnative-pg/actions/runs/9571998035

**Review by gbartolini — APPROVED** [review]

**Review by leonardoce — APPROVED** [review]

**Review by fcanovai — APPROVED** [review]

**Pull request merged**